Tests/FluentKitTests/Utility/DummyDatabaseForTestQueries.swift

```diff
@@ -0,0 +1,97 @@
+@testable import FluentKit
+import NIO
+import SQLKit
+
+/// Lets you determine the row results for each query.
+///
+/// Initialize with an array or arrays, each inner array
+/// containing all the rows for a certain query result. Each
+/// new query will produce the next inner-array of results.
+///
+/// You can specify an empty result-set (as is correct for
+/// some queries and useful for other tests) with an empty
+/// inner-array.
+///
+/// Examples:
+///
+/// Return an empty result for all queries:
+///
+///     DummyDatabaseForTestQueries(mockResults: [[]])
+///
+/// Return an empty result for first query, and a single result
+/// for the second query (perhaps a query to find a record with
+/// no results followed by a successful query to create the record):
+///
+///     DummyDatabaseForTestQueries(mockResults: [
+///         [],
+///         [
+///             DummyDatabaseForTestQueries.DummyRow(dummyDecodedFields: ["id": 1, "name": "Boise"])
+///         ]
+///     ])
+public class DummyDatabaseForTestQueries: Database {
+
+    public typealias MockResult = [DatabaseRow]
+
+    let mockResults: [MockResult]
+    var resultIdx: Int = 0
+
+    public var dialect: SQLDialect {
+        DummyDatabaseDialect()
+    }
+    public var context: DatabaseContext
+
+    public init(
+        mockResults: [MockResult] = [[]],
+        context: DatabaseContext = .init(
+            configuration: .init(),
+            logger: .init(label: "test"),
+            eventLoop: EmbeddedEventLoop()
+        )
+    ) {
+        precondition(mockResults.count > 0, "An empty array of results is not supported. Did you mean to pass 1 empty result (i.e. `[[]]`)?")
+        self.context = context
+        self.mockResults = mockResults
+    }
+
+    public func execute(query: DatabaseQuery, onRow: @escaping (DatabaseRow) -> ()) -> EventLoopFuture<Void> {
+        defer { resultIdx = (resultIdx + 1) % mockResults.count }
```

Comment:
I'm not sure what the benefit of cycling the results would be. I think it would be nice if the results acted like a FIFO queue where you could `append` query results and then each query on the DB popped them from the front of the queue.

Reply:
This was more of a coin flip than anything for me, but I landed on cycling because there did not seem to be any use in the database not responding at all, which would result in test the case hanging (in contrast to responding with an empty result set). For that reason, I thought that running out of mock responses did not serve a purpose and cycling allows you do execute arbitrarily many queries in a loop without worrying about matching the number of mocked responses to the number of expected queries.

Reply:
For loops you could do:

```swift
for x in ... {
    db.append(queryResult: ...)
    Model.query(on: db)...
}
```

I think throwing an error if there's not enough test data would be useful when you're writing tests and something gets off. That seems easier to debug than trying to figure out why one of the tests is getting unexpected data decoding errors.

Reply:
I can get behind that. Done.